esp8266-fastled-webserver/Fields.cpp

```diff
@@ -17,479 +17,526 @@
 */
 #include "common.h"
 
-const String NumberFieldType = "Number";
-const String BooleanFieldType = "Boolean";
-const String SelectFieldType = "Select";
-const String ColorFieldType = "Color";
-const String SectionFieldType = "Section";
-const String StringFieldType = "String";
-const String LabelFieldType = "Label";
-const String UtcOffsetIndexFieldType = "UtcOffsetIndex";
-
-uint8_t power = 1;
-uint8_t brightness = brightnessMap[brightnessIndex];
-
-//String setPower(String value) {
-//  power = value.toInt();
-//  if(power < 0) power = 0;
-//  else if (power > 1) power = 1;
-//  return String(power);
-//}
-
-String getPower() {
-  return String(power);
-}
-
-//String setBrightness(String value) {
-//  brightness = value.toInt();
-//  if(brightness < 0) brightness = 0;
-//  else if (brightness > 255) brightness = 255;
-//  return String(brightness);
-//}
-
-// TODO: wrapper class for EEPROM settings, to help ensure clear understanding
-//       of all locations that can cause settings committed / EEPROM write.
-//       Also simplifies reading prior value before writing new value (extend EEPROM life)
-//       Also simplifies later extending CRC, transactional updates, bypassing "dead"
-//       areas of EEPROM, etc.
-
-String getBrightness() {
-  return String(brightness);
-}
-
-String getPattern() {
-  return String(currentPatternIndex);
-}
-
-String getPatterns() {
-  String json = "";
+// only items outside the anonymous namespace can be exported.
+// this keeps the global namespace cleaner, with no runtime/RAM costs.
+inline namespace {
+  enum struct Field_t : uint8_t {
+    Number,
+    Boolean,
+    Select,
+    Color,
+    Section,
+    String,
+    Label,
+    UtcOffset,
+  };
+  constexpr bool IsValid(const Field_t& t) {
+    return (
+      (t == Field_t::String   ) ||
+      (t == Field_t::Label    ) ||
+      (t == Field_t::Color    ) ||
+      (t == Field_t::Boolean  ) ||
+      (t == Field_t::Number   ) ||
+      (t == Field_t::Section  ) ||
+      (t == Field_t::Select   ) ||
+      (t == Field_t::UtcOffset) );
+  }
 
-  for (uint8_t i = 0; i < patternCount; i++) {
-    json += "\"" + patterns[i].name + "\"";
-    if (i < patternCount - 1)
-      json += ",";
+  const String ToString(const Field_t& t) {
+    return
+      (t == Field_t::String)    ?  F("String")   :
+      (t == Field_t::Label)     ?  F("Label")    :
+      (t == Field_t::Color)     ?  F("Color")    :
+      (t == Field_t::Boolean)   ?  F("Boolean")  :
+      (t == Field_t::Number)    ?  F("Number")   :
+      (t == Field_t::Section)   ?  F("Section")  :
+      (t == Field_t::Select)    ?  F("Select")   :
+      (t == Field_t::UtcOffset) ?  F("UtcOffsetIndex") :
+                                   F("Invalid")  ;
   }
```

Comment:
One could easily live without all these strings by adding (quite) the same enumeration to the **[app.js:](https://github.com/jasoncoon/esp8266-fastled-webserver/blob/b1dd8e9938eca5a93f8d11b1a8bf0389d653b3e6/esp8266-fastled-webserver/data/js/app.js)**
```

var postColorTimer = {};
var postValueTimer = {};

var ignoreColorChange = false;

const fieldtype = Object.freeze({
  Number     : 0,
  Boolean    : 1,
  Select     : 2,
  Color      : 3,
  Section    : 4,
  String     : 5,
  Label      : 6,
  UtcOffset  : 7,
  Invalid    : 8
});

// var ws = new ReconnectingWebSocket("ws://" + address + ":81/", ["arduino"]);

...

$.get(urlBase + "all", function(data) {
      $("#status").html("Loading, please wait...");

      allData = data;

      $.each(data, function(index, field) {
        if (field.type == fieldtype.Number) {
          addNumberField(field);
        } else if (field.type == fieldtype.Boolean) {
          addBooleanField(field);
        } else if (field.type == fieldtype.Select) {
          addSelectField(field);
        } else if (field.type == fieldtype.Color) {
          addColorFieldPalette(field);
          addColorFieldPicker(field);
        } else if (field.type == fieldtype.Select) {
          addSectionField(field);
        } else if (field.type == fieldtype.String) {
          addStringField(field, false);
        } else if (field.type == fieldtype.Label) {
          addStringField(field, true);
        } else if (field.type == fieldtype.UtcOffset) {
          addUtcOffsetIndexField(field, true);
        }
      });

      $(".minicolors").minicolors({
        theme: "bootstrap",
        changeDelay: 200,
        control: "wheel",
        format: "rgb",
        inline: true
      });

...
```

Reply:
Yes, you are right.

This would change the JSON data generated.
I want to avoid that, in case anyone wrote scripts that interpret this JSON.  Changing the underlying data type to an `enum` would break those scripts.  Thus, for at least this PR, I am choosing to avoid breaking changes in the JSON data.